pkg/avx/collector.go

```diff
@@ -71,117 +80,135 @@ var descriptors = [numDescriptors]*prometheus.Desc{
 	}
 
-	currentCode, err := bpf.CurrentKernelVersion()
-	if err != nil {
-		return errors.Wrap(err, "unable to get current kernel version")
+	if _, _, errno := unix.Syscall(unix.SYS_IOCTL, uintptr(pfd), unix.PERF_EVENT_IOC_ENABLE, 0); errno != 0 {
+		return -1, errors.New("unable to set up perf events")
 	}
 
-	if currentCode < kernelVersionCode(elfMajor, elfMinor, 0) {
-		return errors.New("host kernel is too old, consider rebuilding eBPF")
+	if _, _, errno := unix.Syscall(unix.SYS_IOCTL, uintptr(pfd), unix.PERF_EVENT_IOC_SET_BPF, uintptr(prog.FD())); errno != 0 {
+		return -1, errors.New("unable to attach bpf program to perf events")
 	}
 
-	return nil
+	return pfd, nil
 }
 
-type collector struct {
-	root                     string
-	elfFilepath              string
-	bpfModule                *bpf.Module
-	avxContextSwitchCounters *bpf.Map
-	allContextSwitchCounters *bpf.Map
-	lastUpdateNs             *bpf.Map
-	lastCPUCounters          *bpf.Map
-}
+func getKernelVersion() uint32 {
 
-// NewCollector creates new Prometheus collector for AVX metrics
-func NewCollector() (prometheus.Collector, error) {
+	var uts syscall.Utsname
 
-	elfFilepath := path.Join(bpfInstallpath, bpfBinaryName)
+	err := syscall.Uname(&uts)
+	if err != nil {
+		return 0
+	}
 
-	if err := checkElfKernelVersion(elfFilepath); err != nil {
-		return nil, err
+	str := make([]byte, len(uts.Release))
+	for _, val := range uts.Release[:] {
+		if val == 0 {
+			break
+		}
+		str = append(str, byte(val))
 	}
 
-	bpfModule := bpf.NewModule(elfFilepath)
+	ver := strings.SplitN(strings.Trim(string(str), "\x00"), ".", 3)
 
-	sectionParams := make(map[string]bpf.SectionParams)
-	if err := bpfModule.Load(sectionParams); err != nil {
-		return nil, errors.Wrap(err, "unable to load eBPF ELF file")
+	major, err := strconv.ParseUint(ver[0], 10, 8)
+	if err != nil {
+		return 0
+	}
+	minor, err := strconv.ParseUint(ver[1], 10, 8)
+	if err != nil {
+		return 0
 	}
 
-	allSwitchCounters := bpfModule.Map("all_context_switch_count")
-	if allSwitchCounters == nil {
-		return nil, errors.New("map all_context_switch_count not found")
+	// ignore patch version
+	return uint32(major<<16 + minor<<8)
+}
+
+// NewCollector creates new Prometheus collector for AVX metrics
+func NewCollector() (prometheus.Collector, error) {
+
+	// Set rlimit to be able to lock map values in memory
+	memlockLimit := &unix.Rlimit{
+		Cur: mapMemLockLimit,
+		Max: mapMemLockLimit,
+	}
+	err := unix.Setrlimit(unix.RLIMIT_MEMLOCK, memlockLimit)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to set rlimit")
 	}
 
-	avxSwitchCounters := bpfModule.Map("avx_context_switch_count")
-	if avxSwitchCounters == nil {
-		return nil, errors.New("map avx_context_switch_count not found")
+	spec, err := bpf.LoadCollectionSpec(filepath.Join(bpfInstallpath, bpfBinaryName))
+	if err != nil {
+		log.Info("User eBPF not found. Using default CollectionSpec from ELF program bytes")
```

Comment:
Should we print the error (with `%v`) just to be more informative about the error condition?

Reply:
added.